### ISSUE TITLE
Ray Data Q4 Roadmap + Wishlist

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#58665]
### Description

Ray Data community has been rapidly growing every month. We're putting out here the roadmap for the upcoming quarter.

Feel free to leave any requests or questions below. If there's anything missing, please feel free to leave below!

## Read/Write integrations
  * Improved Iceberg Support   
    * Upsert support: 
      * https://github.com/ray-project/ray/issues/53438, 
      * #58270 
    * Overwrite support: 
      * https://github.com/ray-project/ray/issues/51951
    * Leveraging partition / metadata information: 
      * #58286 
    * Creating, altering, deleting tables 
  * Bounded Kafka Support: 
    * https://github.com/ray-project/ray/issues/58653
    * https://github.com/ray-project/ray/pull/58592
  * write_kafka support: 
    * https://github.com/ray-project/ray/issues/58725
  * Turbopuffer Support  
  * Land DataSourceV2
  * Hive support: 
    * https://github.com/ray-project/ray/pull/58253

## Expressions 
  * Exposing core expression ops: https://github.com/ray-project/ray/issues/58674
    * #58465 
    * #58231 
    * #58387

## Query Planning
  * Predicate pushdown: #58150, #58555
  * Projection pushdown: #58151

## Improved streaming support
  * map\_groups to support iterator outputs: #58039

## Preprocessors
  *  Transform optimizations
  *  Migrate Preprocessors to Aggregate V2 : #58371 
  *  Simplify API for custom preprocessors  
  *  Save/Load framework for preprocessors: #58321

## Training Dataloading
  * Improve shuffle flexibility (i.e., shuffle metadata at row groups or row level)
  * Collation off training thread
  * Proper patterns for mixing dataloaders

## Shuffle scalability envelope

We plan to publish a scalability envelope for Ray Data and a roadmap for improving scalability to make Ray Data competitive with other data processing frameworks.

## LLM integrations  
  * Multimodal Data Processors: #58260
  * Stage-based configuration: #58298 
  * Tool-calling: #58091 

## Preview for 2026

* External Shuffle Service: #58687 
* Unbounded data source support: #40513 
* Windowed Expressions  
* Integration with Ibis

**Comment by zzhou612** [comment]
"Integration with Ibis" means that we are planning to add Ray Data as one of the execution backend for Ibis API? (Similiar to Polars Dataframe?) Are we planning to add more dataframe API (similiar to ones in Spark, Polars, Daft) to Ray Data's dataset API?

**Comment by yyzhao2025** [comment]
Could you please elaborate on the background and motivation behind integrating Ibis, whether there is strong user demand for this integration, and if there is already a clear plan in place for the DF API?
